File: atcoder/codefes_2015_yosen_a/d_partial2.rb
def ok(t,n,m,x)
    d=Array.new(m+1,0)
    for i in 1..m
        l=x[i]-d[i-1]-1
        r=0
        while true
            if [l*2+r,r*2+l].min > t || (i!=m && x[i+1] <= x[i]+r)
                if r==0
                    return false
                else
                    r-=1
                    break
                end
            else
                r+=1
            end
        end
        d[i]=x[i]+r
    end
    d[m]>=n
end

n,m=gets.chomp.split.map(&:to_i)
x=Array.new(m+1)

for i in 1..m
    x[i] = gets.to_i
end

lb=0
ub=3*n.to_f

for i in 1..64
    mid = (lb+ub)/2
    if ok(mid,n,m,x)
        ub=mid
    else
        lb=mid
    end
end

puts lb.round
          

File: atcoder/codefes_2015_yosen_a/d_partial2.cpp
#include <bits/stdc++.h>
using namespace std;

int n,m;
vector<int> x,d;

bool ok(double t){
    d[0]=0;
    
    for(int i=1;i<=m;i++){
        int l=x[i-1]-d[i-1]-1,r=0;
        while(true){
            if(min(l*2+r,l+r*2) > t || (i!=m && x[i] <= x[i-1]+r)){
                if(r==0){
                    return false;
                }else{
                    r--;
                    break;
                }
            }else{
                r++;
            }
        }
        d[i]=x[i-1]+r;
    }

    return d[m]>=n;
}

int main(){
    cin>>n>>m;
    d.resize(m+1);
    for(int i=0;i<m;i++){
        int t;
        cin>>t;
        x.push_back(t);
    }

    double lb=0,ub=3*n;

    for(int i=0;i<1000;i++){
        double mid = (lb+ub)/2;

        if(ok(mid)){
            ub=mid;
        }else{
            lb=mid;
        }
    }

    cout << round(lb) << endl;
}
